### ISSUE TITLE
Lua: Interface for moving the cursor to a position

### ISSUE THREAD

**Issue by gpanders** [neovim/neovim#26782]
# Problem

Moving the cursor is a fundamental action in Nvim, and there are many different ways to do so, depending on context:

- `vim.api.nvim_win_set_cursor()` or `vim.fn.cursor()` are the low level "primitives"
- `vim.diagnostic.goto_next()` and `vim.diagnostic.goto_prev()` move between diagnostics
- https://github.com/neovim/neovim/pull/24499 proposed a method for moving between tree-sitter nodes
- `vim.lsp.util.jump_to_location` moves the cursor to a given position within a specific buffer

https://github.com/neovim/neovim/pull/24499#issuecomment-1658752117 mentions an abstract `vim.goto()` or `vim.cursor()` function that accepts any arbitrary "position" and moves the cursor to that location.

Related: https://github.com/neovim/neovim/issues/25509
Related: https://github.com/neovim/neovim/pull/26745

# Proposal

We cannot name the function `vim.goto()` because `goto` is a reserved keyword in LuaJIT. IMO `vim.cursor()` is not a great name because "cursor" is not a verb. So I propose `vim.jump()`: "jump" is a standard and common idiom in Vim terminology and is therefore already familiar with users.

`vim.jump()` accepts a [Position](https://github.com/neovim/neovim/issues/25509) as an argument and moves the cursor to that location (as well as an optional buffer number to support jumping between buffers, to support the LSP use case? Or is the buffer number included in the `Position`?)

`vim.jump()` can then be used for all of the use cases covered in the problem description:

- As the go-to "primitive" for basic cursor movement:

  ```lua
  -- Jump to the beginning of a buffer
  vim.jump({ 0, 0 })

  -- Jump to the last line of a buffer
  vim.jump({ -1, 0 })
  ```

- For use with `vim.diagnostic` and `vim.treesitter` APIs:

  ```lua
  -- Jump to next diagnostic, replaces vim.diagnostic.goto_next()
  vim.jump(vim.diagnostic.next(1))

  -- Jump to previous diagnostic, replaces vim.diagnostic.goto_prev()
  vim.jump(vim.diagnostic.next(-1))

  -- Jump to Tree-sitter node
  local node = vim.treesitter.get_node()
  vim.jump(node:pos())
  ```

- Could possibly be used to replace `vim.lsp.util.jump_to_location`, in service of https://github.com/neovim/neovim/issues/25272

**Comment by clason** [comment]
Just to bring up alternatives: `vim.go_to()` would be legal; `vim.move()` may work, too.

**Comment by clason** [comment]
> vim.jump() accepts a https://github.com/neovim/neovim/issues/25509 as an argument and moves the cursor to that location (as well as an optional buffer number to support jumping between buffers, to support the LSP use case? Or is the buffer number included in the Position?)

I think we touched on optional fields in `vim.pos` (in the context of `byte`, as in `TSRange6`), where bufnr could live as well. But I feel that should be determined by general API considerations ("how often do you need positions in _other_ buffers?")

**Comment by bfredl** [comment]
> We cannot name the function vim.goto() because goto is a reserved keyword in LuaJIT. 

I don't think this is true. it is used as a keyword in statement contexts, but it is not actually reserved (because that would break 5.1 compat).

**Comment by gpanders** [comment]
> > We cannot name the function vim.goto() because goto is a reserved keyword in LuaJIT.
> 
> I don't think this is true. it is used as a keyword in statement contexts, but it is not actually reserved (because that would break 5.1 compat).

Yes, my mistake. It's not LuaJIT that has a problem with `goto()`, but some of our tooling (`luacheck` in particular seems to run into a parse error when using `goto()`). Maybe we can work around this.

**Comment by justinmk** [comment]
`setpos()` is another option.

**Comment by dundargoc** [comment]
> I actually wonder if english grammar ever matters, since we can define anything as a verb/noun/etc: what matters is our internal coherency

I very much think it matters. It's the most obnoxious thing in the world when software create their own cinematic universe and reinvent the dictionary for already well-established terms. It may unnecessarily raise the barrier for use due to the potential confusion it may cause.

**Comment by yorickpeterse** [comment]
> vim.jump() accepts a https://github.com/neovim/neovim/issues/25509 as an argument and moves the cursor to that location (as well as an optional buffer number to support jumping between buffers, to support the LSP use case? Or is the buffer number included in the Position?)

Shouldn't it take a _window_ number instead of a buffer number, given `nvim_win_set_cursor` operates on windows?

As for naming, I would just stick with `vim.cursor` as it makes the most sense/is likely what one would search for when trying to figure out how to move the cursor using Lua. Forcing everything to be verbs doesn't really make sense to me, especially given (at least some) existing modules (e.g. `vim.diagnostic`) don't follow this pattern any way.

**Comment by gpanders** [comment]
> > vim.jump() accepts a #25509 as an argument and moves the cursor to that location (as well as an optional buffer number to support jumping between buffers, to support the LSP use case? Or is the buffer number included in the Position?)
> 
> Shouldn't it take a _window_ number instead of a buffer number, given `nvim_win_set_cursor` operates on windows?

The idea is that this API is a convenient way to move the “active cursor”, so it always operates on the current window. The buffer number argument means that the buffer in the current window changes. The concept is from the point of view of what the user wants to do (“put my cursor in this file on this line and column”). For use cases that need to move the cursor in a different window there is still `nvim_win_set_cursor` and friends. 

> As for naming, I would just stick with vim.cursor as it makes the most sense/is likely what one would search for when trying to figure out how to move the cursor using Lua. Forcing everything to be verbs doesn't really make sense to me, especially given (at least some) existing modules (e.g. vim.diagnostic) don't follow this pattern any way.

Noun-functions imply that it returns a thing. Verb-functions imply that it does a thing. `vim.cursor()` is ambiguous: it could mean that it returns the cursor position, or it could move the cursor. Sticking to verbs is not an ideological point, it’s trying to match user expectations and minimize surprise/ambiguity.

**Comment by dundargoc** [comment]
> especially given (at least some) existing modules (e.g. vim.diagnostic) don't follow this pattern any way.

Wait, maybe I'm outing myself as not knowing english well, but is this even true? These are the following functions from `vim.diagnostic`:

```
get             - verb
set             - verb
hide            - verb
show            - verb
count           - verb
match           - verb
reset           - verb
config          - could argue that this is shorthand for "configure" which would be verb. Could also be noun.
enable          - verb
disable         - verb
get_next        - verb + noun
get_prev        - verb + noun
toqflist        - preposition + noun
goto_next       - verb + noun
goto_prev       - verb + noun
setqflist       - verb + noun
open_float      - verb + noun
fromqflist      - preposition + noun
setloclist      - verb + noun
is_disabled     - verb + noun
get_next_pos    - verb + noun
get_prev_pos    - verb + noun
get_namespace   - verb + noun
get_namespaces  - verb + noun
```

I think most of these are pretty verby tbh.

**Comment by yorickpeterse** [comment]
@gpanders My bad, I thought `vim.cursor` was the _module_ and not a function, i.e. there would be something like `vim.cursor.move`, `vim.cursor.get`, etc.

**Comment by justinmk** [comment]
for reference https://github.com/neovim/neovim/pull/26745#pullrequestreview-2082322699

> > Looks like we've converged on "jump" as a name for navigation, based on `vim.diagnostic.jump()` and `vim.snippet.jump()`.

**Comment by mfussenegger** [comment]
Potential use-case from https://github.com/neovim/neovim/pull/30927 includes updating jumplist/tagstack 

Also a point from recent discussion in chat:
- Might need to support something like a `:getpos()` protocol, so you can compose with cases like `vim.jump(vim.diagnostic.get({ direction = 1 }))`, or `vim.jump(vim.lsp.buf.get_xy())` - for LSP it might be desirable to do lsp.Location offset decoding lazy, and a list might have mixed encodings. But it's unclear how well lsp methods would/will compose, given most are async
